pkg/helm/charts.go

```diff
@@ -0,0 +1,283 @@
+	"helm.sh/helm/v3/pkg/cli"
+	"helm.sh/helm/v3/pkg/getter"
+	"helm.sh/helm/v3/pkg/repo"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+// ProfileAnnotation is the annotation that Helm charts must have to indicate
+// that they provide a Profile.
+const ProfileAnnotation = "weave.works/profile"
+
+// NewRepoManager creates and returns a new RepoManager.
+func NewRepoManager(kc client.Client, cacheDir string) *RepoManager {
+	return &RepoManager{
+		Client:   kc,
+		CacheDir: cacheDir,
+		envSettings: &cli.EnvSettings{
+			Debug:            true,
+			RepositoryCache:  cacheDir,
+			RepositoryConfig: path.Join(cacheDir, "/repository.yaml"),
+		},
+	}
+}
+
+// RepoManager implements HelmRepoManager interface using the Helm library packages.
+type RepoManager struct {
+	client.Client
+	CacheDir    string
+	envSettings *cli.EnvSettings
+}
+
+// ChartReference is a Helm chart reference, the SourceRef is a Flux
+// SourceReference for the Helm chart.
+type ChartReference struct {
+	Chart     string
+	Version   string
+	SourceRef helmv2beta1.CrossNamespaceObjectReference
+}
+
+// DefaultChartGetter provides default ways to get a chart index.yaml based on
+// the URL scheme.
+var defaultChartGetters = getter.Providers{
+	getter.Provider{
+		Schemes: []string{"http", "https"},
+		New:     getter.NewHTTPGetter,
+	},
+}
+
+type ChartPredicate func(*repo.ChartVersion) bool
+
+// Profiles is a predicate for scanning charts with the ProfileAnnotation.
+var Profiles = func(v *repo.ChartVersion) bool {
+	return hasAnnotation(v.Metadata, ProfileAnnotation)
+}
+
+// ScanCharts filters charts using the provided predicate.
+// TODO: Add caching based on the Status Artifact Revision.
+func (h *RepoManager) GetCharts(ctx context.Context, hr *sourcev1beta1.HelmRepository, pred ChartPredicate) ([]*pb.Profile, error) {
+	chartRepo, err := fetchIndexFile(hr.Status.URL)
+	if err != nil {
+		return nil, fmt.Errorf("fetching profiles from HelmRepository %s/%s: %w",
+			hr.GetName(), hr.GetNamespace(), err)
+	}
+
+	ps := make(map[string]*pb.Profile)
+
+	for name, versions := range chartRepo.Entries {
+		for _, v := range versions {
+			if pred(v) {
+				// if already added, update the versions array
+				if p, ok := ps[name]; ok {
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+				} else { // otherwise create a new profile and add to map
+					p = &pb.Profile{
+						Name:        name,
+						Home:        v.Home,
+						Sources:     v.Sources,
+						Description: v.Description,
+						Keywords:    v.Keywords,
+						Icon:        v.Icon,
+						KubeVersion: v.KubeVersion,
+					}
+					for _, m := range v.Maintainers {
+						p.Maintainers = append(p.Maintainers, &pb.Maintainer{
+							Name:  m.Name,
+							Email: m.Email,
+							Url:   m.URL,
+						})
+					}
+					p.AvailableVersions = append(p.AvailableVersions, v.Version)
+					ps[name] = p
+				}
+			}
```

Comment:
```suggestion
			if !pred(v) {
				continue
			}
			// if already added, update the versions array
			if p, ok := ps[name]; ok {
				p.AvailableVersions = append(p.AvailableVersions, v.Version)
			} else { // otherwise create a new profile and add to map
				p = &pb.Profile{
					Name:        name,
					Home:        v.Home,
					Sources:     v.Sources,
					Description: v.Description,
					Keywords:    v.Keywords,
					Icon:        v.Icon,
					KubeVersion: v.KubeVersion,
				}
				for _, m := range v.Maintainers {
					p.Maintainers = append(p.Maintainers, &pb.Maintainer{
						Name:  m.Name,
						Email: m.Email,
						Url:   m.URL,
					})
				}
				p.AvailableVersions = append(p.AvailableVersions, v.Version)
				ps[name] = p
			}
```

Reply:
I am not really a big fan of this :/ I don't mind the nested if. The `continue` is a bit easier to accidentally not see